lib/Interpreter/IncrementalCUDADeviceCompiler.cpp

```diff
@@ -0,0 +1,449 @@
+    if(langOpts.CPlusPlus11)
+      m_CuArgs.cppStdVersion = "-std=c++11";
+    if(langOpts.CPlusPlus14)
+      m_CuArgs.cppStdVersion = "-std=c++14";
+    if(langOpts.CPlusPlus1z)
+      m_CuArgs.cppStdVersion = "-std=c++1z";
+    if(langOpts.CPlusPlus2a)
+      m_CuArgs.cppStdVersion = "-std=c++2a";
+
+    m_CuArgs.optLevel = "-O" + std::to_string(optLevel);
+
+    if(!invocationOptions.CompilerOpts.CUDAGpuArch.empty()){
+      m_CuArgs.ptxSmVersion = "--cuda-gpu-arch="
+                              + invocationOptions.CompilerOpts.CUDAGpuArch;
+      m_CuArgs.fatbinSmVersion = "--image=profile=compute_"
+                              + invocationOptions.CompilerOpts.CUDAGpuArch.substr(3);
+    }
+
+    //The generating of the fatbin file is depend of the architecture of the host.
+    llvm::Triple hostTarget(llvm::sys::getDefaultTargetTriple());
+    m_CuArgs.fatbinArch = hostTarget.isArch64Bit() ? "-64" : "-32";
+
+    m_CuArgs.verbose = invocationOptions.Verbose();
+
+    // FIXME : Should not reduce the fine granulated debug options to a simple.
+    // -g
+    if(debugInfo == clang::codegenoptions::DebugLineTablesOnly ||
+       debugInfo == clang::codegenoptions::LimitedDebugInfo ||
+       debugInfo == clang::codegenoptions::FullDebugInfo)
+      m_CuArgs.debug = true;
+
+    // FIXME : Cling has problems to detect this arguments.
+    /*
+    if(langOpts.CUDADeviceFlushDenormalsToZero)
+      m_CuArgs.additionalPtxOpt.push_back("-fcuda-flush-denormals-to-zero");
+    if(langOpts.CUDADeviceApproxTranscendentals)
+      m_CuArgs.additionalPtxOpt.push_back("-fcuda-approx-transcendentals");
+    if(langOpts.CUDAAllowVariadicFunctions)
+      m_CuArgs.additionalPtxOpt.push_back("-fcuda-allow-variadic-functions");
+    */
+
+    m_CuArgs.fatbinaryOpt = invocationOptions.CompilerOpts.CUDAFatbinaryArgs;
+  }
+
+
+  bool IncrementalCUDADeviceCompiler::generateHelperFiles(){
+    // Generate an empty dummy.cu file.
+    m_DummyCUPath = m_FilePath + "dummy.cu";
+    std::error_code EC;
+    llvm::raw_fd_ostream dummyCU(m_DummyCUPath, EC, llvm::sys::fs::F_Text);
+    if(EC){
+      llvm::errs() << "Could not open " << m_DummyCUPath << ": "
+        << EC.message() << "\n";
+      return false;
+    }
+    dummyCU.close();
+
+    m_PTXFilePath = m_FilePath + "cling.ptx";
+    m_GenericFileName = m_FilePath + "cling";
+    return true;
+  }
+
+  bool IncrementalCUDADeviceCompiler::searchCompilingTools(
+      cling::InvocationOptions & invocationOptions){
+    // Search after clang in the folder of cling.
+    llvm::SmallString<128> cwd;
+    // get folder of the cling executable to find the clang which is contained
+    // in cling
+    // nullptr is ok, if we are the main and not a shared library
+    cwd.append(llvm::sys::path::parent_path(
+      llvm::sys::fs::getMainExecutable(
+        invocationOptions.CompilerOpts.Remaining[0], (void *) &cwd)));
+    cwd.append(llvm::sys::path::get_separator());
+    cwd.append("clang++");
```

Comment:
At the moment, I use cling's clang because it use the same library, like cling. So I can get sure, that cling and clang has the same version. If you mean, it would be better to use the user's clang, because the modifications can cause problems, I will use it. 

I think it is a good idea to add a check function, which compares the clang versions of cling and the user's clang to avoid incompatibility. If the versions are different, I could use cling's clang as backup. Do you agree with me?

Reply:
We know that our patches cause the clang test suite to fail - but unpatched clang causes cling's test suite to fail ;-) so your approach sounds good to me. Can happen in a subsequent PR, though.

Reply:
Ok, I will fix it in a subsequent PR.